src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -0,0 +1,385 @@
+                try {
+                    const membership = new CallMembership(memberEvent, membershipData);
+
+                    if (membership.callId !== "" || membership.scope !== "m.room") {
+                        // for now, just ignore anything that isn't the a room scope call
+                        logger.info(`Ignoring user-scoped call`);
+                        continue;
+                    }
+
+                    if (membership.isExpired()) {
+                        logger.info(
+                            `Ignoring expired device membership ${memberEvent.getSender()}/${membership.deviceId}`,
+                        );
+                        continue;
+                    }
+                    callMemberships.push(membership);
+                } catch (e) {
+                    logger.warn("Couldn't construct call membership: ", e);
+                }
+            }
+        }
+
+        callMemberships.sort((a, b) => a.createdTs() - b.createdTs());
+        logger.debug(
+            "Call memberships, in order: ",
+            callMemberships.map((m) => [m.createdTs(), m.member.userId]),
+        );
+
+        return callMemberships;
+    }
+
+    /**
+     * Return a the MatrixRTC for the room, whether there are currently active members or not
+     */
+    public static roomSessionForRoom(client: MatrixClient, room: Room): MatrixRTCSession {
+        const callMemberships = MatrixRTCSession.callMembershipsForRoom(room);
+
+        return new MatrixRTCSession(client, room, callMemberships);
+    }
+
+    private constructor(
+        private readonly client: MatrixClient,
+        public readonly room: Room,
+        public memberships: CallMembership[],
+    ) {
+        super();
+        this.setExpiryTimer();
+    }
+
+    public isJoined(): boolean {
+        return this.relativeExpiry !== undefined;
+    }
+
+    /**
+     * Performs cleanup & removes timers for client shutdown
+     */
+    public stop(): void {
+        this.leaveRoomSession();
+        if (this.expiryTimeout) {
+            clearTimeout(this.expiryTimeout);
+            this.expiryTimeout = undefined;
+        }
+    }
+
+    /**
+     * Announces this user and device as joined to the MatrixRTC session,
+     * and continues to update the membership event to keep it valid until
+     * leaveRoomSession() is called
+     * This will not subscribe to updates: remember to call subscribe() separately if
+     * desired.
+     */
+    public joinRoomSession(activeFoci: Focus[]): void {
+        if (this.isJoined()) {
+            logger.info(`Already joined to session in room ${this.room.roomId}: ignoring join call`);
+            return;
+        }
+
+        logger.info(`Joining call session in room ${this.room.roomId}`);
+        this.activeFoci = activeFoci;
+        this.relativeExpiry = MEMBERSHIP_EXPIRY_TIME;
+        this.emit(MatrixRTCSessionEvent.JoinStateChanged, true);
+        this.updateCallMembershipEvent();
```

Comment:
This is an `async` function, we don't call `await` here. Is it intentional?

Reply:
It is, I've added a comment. The reason is mostly because it retries, so when `updateCallMembershipEvent()` returns doesn't necessarily mean anything at all it could have just failed to send the event.

Reply:
Yeah, this is true, but I was worried more about is because since we never `await` on it, it's possible to call this function multiple times in a row (essentially making them run in parallel), I don't know if it would have any undesired effects.

I.e. something like:
1. `this.updateCallMembershipEvent()` (gets executed and yields the control flow back to the runtime upon `await` when sending an event / when waiting to I/O).
2. `this.updateCallMembershipEvent()` is called again (while the previous is still running), while reacting to a different event.
3. Possibly `N` instance of `this.updateCallMembershipEvent()` will then run "in parallel".

May be an issue if `this.updateCallMembershipEvent()` relies upon some state that is not meant to be modified "in parallel" (while waiting on the `await`).

Reply:
Good point: awaiting on the funcvtion wouldn't have helped here since other events would still arrive and trigger it while it was awaiting, but I've put a wrapper in that prevents it from running it again if there's already one running.

Reply:
Cool, I think you're referring [to this commit](https://github.com/matrix-org/matrix-js-sdk/pull/3663/commits/268a147a11b3a35e6bb7849d4f67efbd494268dd), right? - If so, are we confident that it's a safe change? - From my perspective it introduces another risk:

1. Something changed -> the `triggerCallMembershipEventUpdate()` is called to force the client to update the state.
2. The update happens, then the function stops at the `await` point awaiting the I/O to complete.
3. Something changed again and the `triggerCallMebershipEventUpdate()` is called again, but since `updateCallMembershipRunning` is `true`, we ignore that attempt to trigger an update and `return` eagerly.
4. Step (2) completes.
5. Due to the early return in step (3), we did not call `updateCallMembershipEvent()` and may have missed another update, the state of the session will be invalid (not synced-up) until the next update.

To me it feels like what we want instead is to schedule the update to happen, but just ensure that they execute one by one. In other words, if one update is running and then an event happens that wants to trigger the call membership event update function, we could either:
1. Execute the function which means that it will run "in parallel" with the already ongoing execution of this same function (that's how it was implemented initially in this PR).
2. Drop the latest request to call this function provided we know that this function is already being executed (that's how it done now with the latest change).
3. Accept the incoming request to call this update function, but don't execute it immediately, let the first function finish and only execute the latest request (that's how I imagined it should have been). There are several ways to implement it, one of which is to introduce some sort of a mutex or async queue. LiveKit JS SDK has some of similar synchronisation primitives, [`Mutex` for instance](https://github.com/livekit/client-sdk-js/blob/8042164c5f29f63e12ed91e97905d5b203088b14/src/room/utils.ts#L427).

Reply:
Yep, good point - we just need to make sure one more update happens however many are triggered, so queueing them feels overkill: I've done a simple boolean flag in 19bf62a62.